### DISCUSSION TITLE
tutorials/scripting/nodes_and_scene_instances

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#92] [User-contributed notes]
# tutorials/scripting/nodes_and_scene_instances

This guide explains how to get nodes, create nodes, add them as a child, and instantiate scenes from code. Getting nodes: You can get a reference to a node by calling the Node.get_node() method. Fo...

https://docs.godotengine.org/en/stable/tutorials/scripting/nodes_and_scene_instances.html

**Comment by wizard7377** [comment]
Note that all nodes will only be available at runtime, therefore, if you set a variable to a node with a node path, it must have the `@onready` attribute

**Comment by guiYudi** [comment]
Worth noting the order in which nodes get loaded and call their `_ready()`, a.k.a. the Tree Order: https://docs.godotengine.org/en/stable/tutorials/scripting/scene_tree.html#tree-order

**Comment by Tiliaya** [comment]
I had an issue where i could move but not rotate an instanced scene from code.
Turns out, i was applying a rotation *before* calling "add_child(instance)", which seemingly reset the rotation but not position.

So always add the instance to the tree before doing anything else with it i guess.

**Reply by Calinou** [reply]
When using physics interpolation, [objects' positions should be set *before* adding them to the scene tree](https://docs.godotengine.org/en/stable/tutorials/physics/interpolation/2d_and_3d_physics_interpolation.html#resetting-physics-interpolation). Otherwise, you'll have to call `Node.reset_physics_interpolation()` on the node afterwards, which is wasteful in terms of performance.

Setting properties on a node post-instantiation also has various other ramifications (especially performance-related), as some operations will essentially have to run twice on initialization rather than just once. This can have a significant cost if you're instantiating dozens of nodes in a single frame.

**Comment by Superbloxen** [comment]
Question: Is it possible to manipulate an instantiated object from the script it was created from? I would like to do this to cause instantiated objects to spawn in front of the player

**Reply by marcinsendyka** [reply]
Yes, it's possible, please see following snippet:

```gdscript

func _ready() -> void:
	var ball_instance = ball_scene.instantiate()
	add_child(ball_instance)
	ball_instance.global_position = Vector2(
		get_viewport().size.x / 2,
		get_viewport().size.y / 2
	)
```
In this case, newly instantiated object is set to be in the middle of the viewport, but you can use other node's (player in your example) position or other calculated value

**Comment by llukehaskell** [comment]
What is the difference between doing the .load("path") and doing

@export var example_scene: PackedScene

and then populating the scene in the Inspector panel by dragging the appropriate .tscn file on to it?

**Reply by SchusJon13** [reply]
Its more intuitive

**Comment by andrea-feather** [comment]
Note that `_Ready()` is not called on nodes from instantiated scenes until they're added to the main scene! You have to add them to the scene tree with `AddChild()` (or similar) or manually call `_Ready()`, otherwise your initialization code will never be called.

**Comment by marcinsendyka** [comment]
Is there recommended way to setup properties for instantiated scene? like having a constructor or init method? I can set all parameters all by one, but that's easy to forget and hard to maintain (after adding new parameter, need to find all usages manually etc.)

```gdscript
var score: LeaderboardScore = score_scene.instantiate()
# setting properties one by one, manually
score.player = "player"
score.score = 5.0
```

**Reply by urs4n** [reply]
Depending on your use case, you could just set default values for the properties of the instantiated scene, but that means that every instance will have the same starting values and you will still need to adjust them one by one if you need to. For example:
```gdscript
extends Node
class_name LeaderboardScore

player: String = "player"
score: float = 5.0
```

Alternatively, you could define a [resource](https://docs.godotengine.org/en/stable/tutorials/scripting/resources.html) that holds the data for you scene and pass that to your newly created instance. The advantage is that you can still set default values like above, but there is only one container for the data that needs to be grouped and it can drive whatever else needs it without having to copy-paste its content around. It's like passing a `struct` directly instead of passing the values inside your `struct` one-by-one.

```gdscript
extends Resource
class_name ScoreData

player: String
score: float = 0.0

# --- In some gameplay-related script

player1_score = ScoreData.new()
player1_score.player = "player 1" # the `score` property starts at 0.0 because we set its default value

player2_score = ScoreData.new()
player2_score.player = "player 2" # the `score` property starts at 0.0 because we set its default value

# ... Update score as the game unfolds ...

player1_score.score += 1.0
player2_score.score += 2.0

# ... When you need to create your score scene
var score: LeaderboardScore = score_scene.isntantiate()
score.data = player1_score # All the properties are set and can drive whatever logic you need
```
